File: dotnet/UpsChecklist.Core/EvidenceValidation.cs
using System.Text.RegularExpressions;

namespace UpsChecklist.Core;

public static partial class EvidenceValidation
{
    private const int MaxLength = 800_000;

    [GeneratedRegex(@"^data:image/(jpeg|png);base64,([A-Za-z0-9+/]+={0,2})$")]
    private static partial Regex DataUrlPattern();

    public static bool ValidEvidencePhoto(string value)
    {
        if (string.IsNullOrEmpty(value))
            return true;
        if (value.Length > MaxLength)
            return false;

        var match = DataUrlPattern().Match(value);
        if (!match.Success)
            return false;

        try
        {
            var bytes = Convert.FromBase64String(match.Groups[2].Value);
            return match.Groups[1].Value == "png"
                ? ValidPng(bytes)
                : ValidJpeg(bytes);
        }
        catch
        {
            return false;
        }
    }

    private static bool ValidSize(int width, int height) =>
        width > 0 && height > 0 && width <= 1600 && height <= 1600 && width * height <= 2_000_000;

    private static bool ValidPng(byte[] bytes)
    {
        ReadOnlySpan<byte> signature = [137, 80, 78, 71, 13, 10, 26, 10];
        if (bytes.Length < 33 || !bytes.AsSpan(0, 8).SequenceEqual(signature))
            return false;

        if (ReadUInt32Be(bytes, 8) != 13)
            return false;
        if (ReadUInt32Be(bytes, 12) != 0x49484452)
            return false;

        var width = (int)ReadUInt32Be(bytes, 16);
        var height = (int)ReadUInt32Be(bytes, 20);
        return ValidSize(width, height);
    }

    private static uint ReadUInt32Be(byte[] bytes, int offset) =>
        ((uint)bytes[offset] << 24) | ((uint)bytes[offset + 1] << 16) | ((uint)bytes[offset + 2] << 8) | bytes[offset + 3];

    private static bool ValidJpeg(byte[] bytes)
    {
        if (bytes.Length < 4 || bytes[0] != 0xFF || bytes[1] != 0xD8)
            return false;

        var offset = 2;
        while (offset < bytes.Length)
        {
            if (bytes[offset] != 0xFF)
                return false;
            while (offset < bytes.Length && bytes[offset] == 0xFF)
                offset++;
            if (offset >= bytes.Length)
                return false;

            var marker = bytes[offset++];
            if (marker is 0xDA or 0xD9)
                return false;
            if (marker is 1 or >= 0xD0 and <= 0xD8)
                continue;

            if (offset + 1 >= bytes.Length)
                return false;
            var length = (bytes[offset] << 8) | bytes[offset + 1];
            if (length < 2 || offset + length > bytes.Length)
                return false;

            if (marker is 0xC0 or 0xC1 or 0xC2)
            {
                if (length < 8)
                    return false;
                var height = (bytes[offset + 3] << 8) | bytes[offset + 4];
                var width = (bytes[offset + 5] << 8) | bytes[offset + 6];
                return ValidSize(width, height);
            }

            offset += length;
        }

        return false;
    }
}

File: lib/evidence-validation.ts
const validSize = (width: number, height: number) => width > 0 && height > 0 && width <= 1600 && height <= 1600 && width * height <= 2_000_000;

export function validEvidencePhoto(value: string): boolean {
  if (value === "") return true;
  if (value.length > 800000) return false;
  const match = /^data:image\/(jpeg|png);base64,([A-Za-z0-9+/]+={0,2})$/.exec(value);
  if (!match) return false;
  try {
    const bytes = Uint8Array.from(atob(match[2]), c => c.charCodeAt(0));
    const view = new DataView(bytes.buffer);
    if (match[1] === "png") {
      const signature = [137, 80, 78, 71, 13, 10, 26, 10];
      if (bytes.length < 33 || !signature.every((byte, i) => bytes[i] === byte) || view.getUint32(8) !== 13 || view.getUint32(12) !== 0x49484452) return false;
      return validSize(view.getUint32(16), view.getUint32(20));
    }
    if (bytes[0] !== 255 || bytes[1] !== 216) return false;
    for (let offset = 2; offset < bytes.length;) {
      if (bytes[offset] !== 255) return false;
      while (bytes[offset] === 255) offset++;
      const marker = bytes[offset++];
      if (marker === 0xda || marker === 0xd9) return false;
      if (marker === 1 || (marker >= 0xd0 && marker <= 0xd8)) continue;
      const length = view.getUint16(offset);
      if (length < 2 || offset + length > bytes.length) return false;
      if ([0xc0, 0xc1, 0xc2].includes(marker)) {
        return length >= 8 && validSize(view.getUint16(offset + 5), view.getUint16(offset + 3));
      }
      offset += length;
    }
  } catch { return false; }
  return false;
}
